### PR TITLE
File descriptor flag support

### PR DESCRIPTION
**Pull request by marcbachmann**
Not sure whether it would be simpler passing just a fourth arg 😄 
The backwards compatibility is probably also not needed.

This pr
- changes the arguments to an object
- adds a `flags` param
- adds a `cache` param

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -4,16 +4,25 @@ var Cache = require('hashlru')
...
-    file
-  )
+module.exports = function (opts) {
+  if (typeof opts !== 'object') opts = legacy(arguments[0], arguments[1], arguments[2])
+
+  var file = opts.file
+  var blockSize = opts.blockSize || 1024*16
+  var codec = opts.codec || {encode: id, decode: id}
+  var flags = opts.flags || 'a+'
+  var cache = opts.cache || Cache(1024)
```
Comment (marcbachmann) [inline index.js]: We could also remove the cache fallback as there is one in the `aligned-block-file` module

**Comment by marcbachmann** [comment]
I'm currently experimenting with the flags.
Maybe it's better to explicly set read & write mode instead of passing the flags.

**Comment by marcbachmann** [comment]
I wonder why we have to use truncate instead of just resetting the in the blocks file. https://github.com/flumedb/flumelog-offset/blame/master/frame/recoverable.js#L54-L68

That way it would read-only compatible and also restorable from there.

**Comment by marcbachmann** [comment]
The flag itself is really usable. But additionally I was trying to refresh an already opened file. That didn't work because of the truncate and the other process which was writing to the file.

**Comment by marcbachmann** [comment]
I slightly changed the args again to use the filename as first arg and the options as second.
That way it's similar to the api of `fs.createReadStream` and `pull-file`

File: index.js [hunk 2]
```diff
@@ -4,16 +4,25 @@ var Cache = require('hashlru')
...
-  var blocks = Blocks(file, block_size,'a+', Cache(1024))
-  return inject(
-    blocks,
-    createFrame(blocks, codec),
-    codec,
-    file
-  )
+module.exports = function (file, opts) {
+  if (!opts) opts = {}
+  if (typeof opts !== 'object') legacy(opts, arguments[2])
```
Comment (marcbachmann) [inline index.js]: Damn, this is still wrong. Opts could be a codec.

**Comment by dominictarr** [comment]
no truncate: https://github.com/flumedb/flumelog-offset/pull/4#issuecomment-311819374

That is quite a good idea... hmm, it may mean that `a`ppend mode doesn't work, though. It could just check if it's read only, set the offset and skip truncating. are you expecting to read from a file that is being appended to by someone else? cos that will have other edge cases.

**Comment by dominictarr** [comment]
thanks, merged into 3.1.0

**Comment by marcbachmann** [comment]
Hmm. we still need to update the arg check in https://github.com/flumedb/flumelog-offset/pull/4#discussion_r125148675

`typeof obj !== 'object' || obj.decode` would do it

> are you expecting to read from a file that is being appended to by someone else? cos that will have other edge cases.

Yes. I'm currently writing my own read-only version. I don't need all the caching in the block-file and reads are already optimized. We might need to modularize the module a bit.

e.g.
```
pull(
  file(`${__dirname}/index`, {live: true}),
  pull.map(decode(codec)),
  pull.log()
)
```

**Comment by marcbachmann** [comment]
I'll write views in a separate process to ensure uptime and reliabilty of the first process.
Views not needed to ensure writes aka indexes will stay in the main process.

**Comment by marcbachmann** [comment]
Next time I should add a WIP annotation somewhere. 😄
Fix in #5

**Comment by marcbachmann** [comment]
The pull stream version also doesn't work of course because it continue if the last block gets written. So I'll use pub/sub on the master process.

Skipping the truncate function would be nice if `flags[0] == 'r'`.
I'll keep you updated.

**Pull request merged**